### ISSUE TITLE
Allow types (when used in method and function) to be more than 1 type.

### ISSUE THREAD

**Issue by rquadling** [docbook/docbook#91]
Hi.

Coming from PHP, specifically PHP 7.1 where the concept of union types are supported.

Currently, where a method may return more than 1 type - commonly a boolean false and some other type, this is documented as `mixed`.

For example:

`mixed strpos ( string $haystack , mixed $needle [, int $offset = 0 ] )`

This is rendered from this docbook snippet:
```xml
  <methodsynopsis>
   <type>mixed</type><methodname>strpos</methodname>
   <methodparam><type>string</type><parameter>haystack</parameter></methodparam>
   <methodparam><type>mixed</type><parameter>needle</parameter></methodparam>
   <methodparam choice="opt"><type>int</type><parameter>offset</parameter><initializer>0</initializer></methodparam>
  </methodsynopsis>
```

The result for this function will either be a boolean `false` or a positive integer, but as a single `<type>` is allowed, `<type>mixed</type>` is what we are currently using.

It would be useful to enhance docbook to allow for multiple types for `<methodsynopsis>` and `<methodparam>`.

So, my proposal is a new tag of `<types>` that allows for 2 or more `<type>`.

```
methodsynopsis ::=
	Sequence of:
		Zero or more of:
			modifier
		Optionally one of:
			type
			types
				Two or more of:
					type
			void
		methodname
		One of:
			One or more of:
				group (db.group.methodparam)
				methodparam
			void
		Zero or more of:
			exceptionname
		Zero or more of:
			modifier
```

and

```
methodparam ::=
	Sequence of:
		Zero or more of:
			modifier
		Optionally one of
			type
			types
				Two or more of:
					type
		One of:
			Sequence of:
				Zero or more of:
					modifier
				parameter
				initializer?
			funcparams
		Zero or more of:
			modifier
```

If there are other places where a `<type>` is defined, these may also need to be expanded to allow for multiple types.

If there is no way to require 2 or more of a child tag, then 1 or more would be fine, and would make `<type>int</type>` the same as `<types><type>int</type></types>`.

I'm happy to make a proper pull request if pointed in the right direction.

**Comment by rquadling** [comment]
Also, the `<type>` for `<methodparam>` would obey the same rules as I've suggested. If there are other places where a `<type>` is defined, these may also need to be expanded to allow for multiple types.

**Comment by bobstayton** [comment]
Hi Richard,
I see you filed this as an issue on the DocBook Github site, so it will 
be considered by the DocBook Technical Committee at our next meeting on 
May 17.

Bob Stayton
Sagehill Enterprises
bobs@sagehill.net

**Comment by rquadling** [comment]
Thank you.

**Comment by rquadling** [comment]
What was the outcome? Are the minutes available?

**Comment by rquadling** [comment]
Just for anyone reading this ...

https://lists.oasis-open.org/archives/docbook-tc/201705/msg00008.html
https://lists.oasis-open.org/archives/docbook-tc/201706/msg00003.html

**Comment by rquadling** [comment]
https://lists.oasis-open.org/archives/docbook-tc/201707/msg00003.html

With regards to the comment ...

> I am not sure that the "types" wrapper is needed or whether you could just allow multiple type elements ...

My reasoning is that it is slightly easier to examine a node to see how many child nodes. Not a great reason maybe, just a little easier to parse, knowing `<types>` only ever contains `<type>`s.

Also, by introducing `<types>`, you aren't breaking backward compatibility. Currently, only 1 of `<type>` or `<void>` is allowed. http://tdg.docbook.org/tdg/5.0/methodsynopsis.html

```
Optionally one of:
  type
  void
```

By introducing a new element `<types>`, you are still allowing optionally 1 of `<type>` or `<void>` or `<type>`. Without `<types>`, you are changing `<type>`s restriction to something like:

```
Optionally one of:
  One or more of:
    type
  void
```

Still maybe not the best of reasons. Hopefully though enough to influence your decision to add them.

**Comment by LarryRowland** [comment]
I agree that adding multiple types is quite a good idea.  I do a lot of Python work and find that being able to return types that are appropriate for different conditions useful. 

I am not entirely clear how allowing multiple `<type>` elements as siblings without a wrapper element would make it incompatible with the current model.  A single `<type>` element is still allowed, as is a single `<void>` element.  That should mean that any current content  created using the existing schema is still allowed.

In terms of interpretation, since it is a sequence, all the `<type>` elements would be visually adjacent and should be easily recognized.  An XPATH `count` function will return zero if no `<type>` element is present, one for a single and N>1 for multiple.  Would that not be sufficient to handle rendering considerations?

I am hesitant to add a wrapper element that is almost the same as a child element that simply allows multiple copies of the child element, particularly since it has no real semantic meaning that is different from simply allowing the multiple `<type>` elements.  Am I missing something here?

**Comment by rquadling** [comment]
Whilst I like the wrapper, I can see that it doesn't offer anything semantically. If this is the only sticking point then I'm happy to do away with it. The main requirement is to allow for multiple types for parameters to methods/functions as well as return types.

**Comment by kosek** [comment]
Beta version of schema with this change is available at

http://docbook.org/xml/5.2b03/

Please give it a try. Any feedback is welcomed.

**Comment by rquadling** [comment]
Thank you. When/if I get any feedback from this, I bring it here.

Based upon the changes made in https://github.com/docbook/docbook/pull/96, I'd consider this issue closed.

Thank you for your consideration.

**Comment by morrisonlevi** [comment]
Coming to this late. My concern is that multiple type declarations does not inherently mean a union; consider that Ceylon has intersection types as well as unions. Having multiple `type` elements in a row wouldn't necessarily mean a union.

**Comment by kosek** [comment]
> Coming to this late. My concern is that multiple type declarations does not inherently mean a union; consider that Ceylon has intersection types as well as unions. Having multiple type elements in a row wouldn't necessarily mean a union.

So far there is not attached any special semantic to multiple type. Could you provide us with concrete examples of existing documentation where there are multiple types meaning intersection and where there is union? It's possible to model this even without wrapper. But we need to know requirements first.

**Comment by rquadling** [comment]
https://ceylon-lang.org/documentation/1.3/tour/types/ Documents intersection and union types.

The use case I came here with was only union types.

Currently, PHP doesn't have any direct support for intersection types (an interesting idea that could be implemented in userland code).

**Comment by kosek** [comment]
Are these intersection types really widely used and is there need to have support for such specific feature in DocBook?

I think that if anyone needs intersection types he/she can easily do it just by adding role="intersection" to all type elements to differentiate this from union of types.

If one day this becomes very common, we can add this natively into DocBook perhaps by introducing class attribute.

**Comment by kosek** [comment]
> TypeScript and Flow also have intersection types and Scala is getting them in 3.0. These are languages in use today (well, Scala 3.0 is coming soon).

And is anyone using DocBook for documenting TS/Flow/Scala libraries? If no one is going to use feature then it's not worth introducing complexity into schema and processing tools.

> Therefore it seems prudent to require something such as role in all cases. I think relation /rel is more meaningful for the attribute name, but I'm not sure how it is decided to create a new attribute compared to just using the generic role.

It doesn't make sense to require such indication on every occurence of type element. The most common case can be default (perhaps union) or this can be completely unspecified if there will be only single type element.

If there is real demand for distinguishing between union/intersection then standard naming convention used elsewhere should be followed and attribute should be named class. E.g.

````
<type class="intersection">xxx</type>
<type class="intersection">yyy</type>
````


````
<type class="union">xxx</type>
<type class="union">yyy</type>
````

**Comment by morrisonlevi** [comment]
Union and intersection are binary; it does not make any sense to put the `class` or `role` on each type. It only makes sense on an aggregate:

    <types relation="union"><type>xxx</type><type>yyy</type></types>

Make sense?

**Comment by kosek** [comment]
Not to me. DocBook is not type modelling language. It should stay generic enough. Previously it was agreed there is no need for wrapping element like `types`. Without wrapper the only sensible place to place such information is type element itself. Yes it's little bit redundant, but such code is usually autogenerated from actual API anyway.

Even if decide there should be wrapper then more semantically correct way would be to use something like:

````
<type class="union">
  <type>xxx</type>
  <type>yyy</type>
</type>
````

Because union type is also type ;-)

**Comment by kosek** [comment]
Implemented in 5.2b05
Schema is available for testing at https://docbook.org/xml/5.2b05/
